web/modules/custom/joinup_licence/src/CompatibilityDocumentListBuilder.php

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * Provides a list controller for the compatibility document entity type.
+ */
+class CompatibilityDocumentListBuilder extends EntityListBuilder {
+
+  /**
+   * The redirect destination service.
+   *
+   * @var \Drupal\Core\Routing\RedirectDestinationInterface
+   */
+  protected $redirectDestination;
+
+  /**
+   * The licence compatibility plugin manager.
+   *
+   * @var \Drupal\joinup_licence\JoinupLicenceCompatibilityRulePluginManager
+   */
+  protected $pluginManager;
+
+  /**
+   * Constructs a new CompatibilityDocumentListBuilder object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type definition.
+   * @param \Drupal\Core\Entity\EntityStorageInterface $storage
+   *   The entity storage class.
+   * @param \Drupal\Core\Routing\RedirectDestinationInterface $redirect_destination
+   *   The redirect destination service.
+   * @param \Drupal\joinup_licence\JoinupLicenceCompatibilityRulePluginManager $plugin_manager
+   *   The licence compatibility rule plugin manager.
+   */
+  public function __construct(EntityTypeInterface $entity_type, EntityStorageInterface $storage, RedirectDestinationInterface $redirect_destination, JoinupLicenceCompatibilityRulePluginManager $plugin_manager) {
+    parent::__construct($entity_type, $storage);
+    $this->redirectDestination = $redirect_destination;
+    $this->pluginManager = $plugin_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity_type.manager')->getStorage($entity_type->id()),
+      $container->get('redirect.destination'),
+      $container->get('plugin.manager.joinup_licence_compatibility_rule')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function render() {
+    $this->ensureEntitiesExist();
+
+    return ['table' => parent::render()];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildHeader() {
+    $header['id'] = $this->t('ID');
+    return $header + parent::buildHeader();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildRow(EntityInterface $entity) {
+    /* @var $entity \Drupal\joinup_licence\Entity\CompatibilityDocumentInterface */
+    $row['id'] = $entity->id();
+    return $row + parent::buildRow($entity);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getDefaultOperations(EntityInterface $entity) {
+    $operations = parent::getDefaultOperations($entity);
+    $destination = $this->redirectDestination->getAsArray();
+    foreach ($operations as $key => $operation) {
+      $operations[$key]['query'] = $destination;
+    }
+    return $operations;
+  }
+
+  /**
+   * Creates missing compatibility documents.
+   *
+   * There should be one compatibility document for each compatibility rule.
+   */
+  protected function ensureEntitiesExist(): void {
```

Comment:
What happens when a plugin is removed from the code? Or when a 3rd-party, shipping such a plugin is uninstalled?

Reply:
We will have to deal with this situation when a plugin is removed from the code.

Reply:
But it's so simple to do this in this method. We just find all entities w/o a plugin and remove them.

Reply:
Actually,

```php
$all_plugins = ...
$all_entities = ...
$entities_to_create = array_diff($all_plugins, $all_entities);
$entities_to_remove = array_diff($all_entities, $all_plugins);
// Action!
```